### ISSUE TITLE
Translations are not ordered in a helpful way

### ISSUE THREAD

**Issue by trang** [Tatoeba/tatoeba2#2364]
Translations are currently ordered alphabetically by language name. When a sentence has a lot of translations and the user is interested in languages that are at the bottom of the list, they have to expand the translations and scroll a lot.

This issue has been intensified with the [transition towards the new sentence design](https://tatoeba.org/eng/wall/show_message/35389#message_35389) because each translation takes twice the space vertically.

Related: #2354

**Comment by alanfgh** [comment]
Is there any chance that this can be implemented before the old sentence design is eliminated? Has there been a discussion of possible approaches? How much effort would be required?

I guess the following existing data could be taken into account:
- the user's profile languages
- the list of languages under Settings -> Options

But this might require some discussion, since the list of languages under Settings -> Options is currently used as a filter (only sentences in languages that appear in the list are displayed), and here we'd be using it as a sorting guide (translations in languages that appear in the list are displayed at the top of the lists of translations, but translations in other languages are displayed as well).

**Comment by trang** [comment]
> Is there any chance that this can be implemented before the old sentence design is eliminated?

Yes. Keep in mind that the old design won't disappear as soon as the new design becomes the default one. It will still be possible to revert to the old design for a certain period of time after the new design has become the default one. 

I don't think we will implement this before we set the new design as default, but there's definitely a chance that this gets implemented before we remove the old design completely.

> Has there been a discussion of possible approaches?

There hasn't yet, but as with many features, we will implement something, let the community test, then readjust the implementation based on the feedback.

The most basic implementation would be to reuse the languages in the Settings and add an option "Show translations only in my languages". This option would be true by default. People who are currently using the languages option would then have no impact. People who want only to prioritize languages but still see translations in all languages would uncheck this option.

Alternatively, if we implement #263, we may not need to have this extra option. We could consider always hiding translations in other languages and whenever the user wants to see translations in all languages, they would use the button to show all translations.

> How much effort would be required?

The most basic implementation isn't very high effort. It's basically 3 things:
- Adding a new option in the settings.
- Taking into account this option when fetching translations.
- Sorting the translations based on the order of the languages in the settings.

**Comment by alanfgh** [comment]
That all sounds fine to me. There's another situation to consider: the case where the user has specified that they want to see only sentences with translations in language X, but hasn't limited display of the translations to language X. In this case, it seems pretty clear to me that translations in language X should be listed first. Currently, they are not.

**Comment by gioele-santi** [comment]
I started checking the code for this issue.
As far as I understand to add the sorting option, we should edit `UsersTable.php`, right?

I am also looking for the code that loads translations.
I started by checking `SentencesController.php` .  By checking its `show` function I cannot find a precise reference to the term **translation** but I wonder if they may be connected to  `Sentences->getContributionsRelatedToSentence($id)` or `getListsForSentence($id)`.

**Comment by AndiPersti** [comment]
> As far as I understand to add the sorting option, we should edit `UsersTable.php`, right?

For adding a new setting you need to update `$defaultSettings` in `Model/Entity/User.php`.

> I am also looking for the code that loads translations.

In the controller, `Sentences->getSentenceWith()` will load the sentence and all the additional data including translations: 
https://github.com/Tatoeba/tatoeba2/blob/b0d88d8a9fc4a2121a55faec473ee5f0bea4903a/src/Controller/SentencesController.php#L168-L169

This works by ["eagerly loading the associations via contain"](https://book.cakephp.org/3/en/orm/retrieving-data-and-resultsets.html#eager-loading-associations-via-contain), i.e. `Sentences` and `Translations` are [linked together](https://github.com/Tatoeba/tatoeba2/blob/b0d88d8a9fc4a2121a55faec473ee5f0bea4903a/src/Model/Table/SentencesTable.php#L67-L72) using a [`BelongsToMany` association](https://book.cakephp.org/3/en/orm/associations.html#belongstomany-associations).

**Comment by gioele-santi** [comment]
I still have to figure out a couple of things, but I would like to work on this.

**Comment by jiru** [comment]
@gioele-santi You are welcome to try working on this. :slightly_smiling_face: I think you can do it like @trang said. There shouldn’t be a lot of code change involved, but it might be a bit tricky.

Here is a technical tip: it’s possible to sort against a list of provided values with the [SQL trick `ORDER BY FIND_IN_SET(...)`](https://www.tutorialspoint.com/mysql-order-by-with-custom-field-value). We use this trick already to [sort search results as returned by the search engine](https://github.com/Tatoeba/tatoeba2/blob/58ad65f02cfd06c3d3ccb155dc9632fec6317668/src/Model/Behavior/SphinxBehavior.php#L162).
